### ISSUE TITLE
Make a clearer distinction between resolved and unresolved types

### ISSUE THREAD

**Issue by Techcable** [DuckLogic/Ivan#1]
Right now we use the same representation for an unknown type that comes out of the parser and for code generation. This is bad.

We should have a clearer phase where we resolve abstract names like "Basic" to actually known definitions. We should probably get rid of that "TypeContext" thing that changes stuff from "UnresolvedTypeRef" to the resolved counterparts "InterfaceTypeRef" and "OpaqueTypeRef".

Basically resolved/unresolved should be part of the type system - not implicit in the state.
